ReplicaEngine/tla/ReplicaEngine.tla

```diff
@@ -0,0 +1,834 @@
+        await versionMap_needsSafeAccess = FALSE;
+        versionMap_isUnsafe := TRUE;
+    end while;
+end process;
+
+(* Other concurrent activity can increase the maxUnsafeAutoIdTimestamp *)
+process MaxUnsafeAutoIdTimestampIncreaserProcess = "MaxUnsafeAutoIdTimestampIncreaser"
+begin
+    MaxUnsafeAutoIdTimestampIncreaserLoop:
+    while pc["Consumer"] /= "Done" do
+        with newTimestamp \in {DocAutoIdTimestamp - 1, DocAutoIdTimestamp, DocAutoIdTimestamp + 1} do
+            await maxUnsafeAutoIdTimestamp < newTimestamp;
+            maxUnsafeAutoIdTimestamp := newTimestamp;
+        end with;
+    end while;
+end process;
+
+(* Lucene refreshes can happen at any time *)
+process LuceneProcess = "ReplicaLucene"
+variables
+    lucene_document = NULL,
+    lucene_buffer   = <<>>,
+begin
+    LuceneLoop:
+    while pc["Consumer"] /= "Done" \/ lucene_buffer /= <<>> do
+
+        lucene_document := ApplyBufferedOperations(lucene_buffer, lucene_document);
+        lucene_buffer := <<>>;
+            
+        (* TODO Model the inner structure of the version map so this refresh can be
+        broken into the individual steps that occur concurrently with ongoing indexing. *)
+                        
+        versionMap_isUnsafe := FALSE;
+        versionMap_needsSafeAccess := FALSE;
+        
+        if versionMap_entry /= NULL
+        then
+            if versionMap_entry.type = UPDATE
+            then
+                versionMap_entry := NULL;
+            else
+                assert versionMap_entry.type = DELETE;
+                versionMap_entry := [ versionMap_entry EXCEPT !.flushed = TRUE ];
+            end if;
+        end if;           
+    end while;
+end process;
+
+(* Flushed deletes expire after a time and are cleaned up *)
+process DeleteCollectorProcess = "DeleteCollector"
+begin
+    DeleteCollectorLoop:
+    while pc["Consumer"] /= "Done" do
+        await /\ versionMap_entry /= NULL
+              /\ versionMap_entry.type = DELETE
+              /\ versionMap_entry.seqno <= localCheckPoint \* PR #28790
+              /\ versionMap_entry.flushed = TRUE;
+        versionMap_entry := NULL;
+    end while;
+end process;
+
+(* Local checkpoint advances as each operation is marked as completed *)
+process LocalCheckpointTrackerProcess = "LocalCheckpointTracker"
+variables
+    localCheckPoint = 0,
+    completedSeqnos = {}
+begin
+    LocalCheckpointTrackerLoop:
+    while pc["Consumer"] /= "Done" do
+        await localCheckPoint + 1 \in completedSeqnos;
+        localCheckPoint := localCheckPoint + 1;
+    end while;
+end process
+
+(* The process that consumes replication requests for a particular document ID, which
+   are processed in series because of the lock in the version map. *)
+process ConsumerProcess = "Consumer"
+variables
+    maxUnsafeAutoIdTimestamp \in {0, DocAutoIdTimestamp - 1, DocAutoIdTimestamp, DocAutoIdTimestamp + 1},
+    maxSeqNoOfNonAppendOnlyOperations \in {0, 2, request_count + 1},
```

Comment:
why only `0, 2, request_count + 1`? How about starting with 0, and then (similarly as you have done for modeling other concurrent requests for different documents) model a separate process that increases this.

Reply:
I was battling state space explosion. However, this doesn't seem to be the source of it, so I pushed 36454d7dc56c6ff9c90332f7b2ede0c729cf43f7.